### PR TITLE
feat: add k-nearest neighbors algorithm

### PR DESCRIPTION
**Pull request by luizcarloscf**
#### Description of Change

* Add [K-nearest neighbors algorithm](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm), a supervised machine learning method used for classification. Using [euclidean distance](https://en.wikipedia.org/wiki/Euclidean_distance) as distance metric between samples.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: machine_learning/k_nearest_neighbors.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+/**
+ * @namespace machine_learning
+ * @brief Machine learning algorithms
+ */
+namespace machine_learning {
+
+/**
+ * @namespace k_nearest_neighbors
+ * @brief K-nearest neighbors algorithm
```
Comment (Panquesito7) [inline machine_learning/k_nearest_neighbors.cpp:28]:
```suggestion
 * @brief Functions for the [K-Nearest Neighbors algorithm]
 * (https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm) implementation
```

**Review by Panquesito7 — DISMISSED** [review]

**Review by Panquesito7 — APPROVED** [review]
Amazing work! Thank you so much. 🚀

File: machine_learning/k_nearest_neighbors.cpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+    /**
+     * Move assignment operator for class Knn
+     */
+    Knn& operator=(Knn&&) = default;
+
+    /**
+     * @brief Destroy the Knn object
+     */
+    ~Knn() = default;
```
Comment (krishnacx) [inline machine_learning/k_nearest_neighbors.cpp:95]: Super

**Comment by luizcarloscf** [comment]
Hey @Panquesito7 ,

I was looking at the organization's repositories and had a question. Some machine learning algorithms can be used for both classification and regression tasks, such as K-Nearest Neighbors or Decision Trees. At [TheAlgorithms/Python](https://github.com/TheAlgorithms/Python/tree/master/machine_learning), there seems to be some distinction in the filenames, but not very clear.

That said, it would make sense to rename the file here like `k_nearest_neighbors_classifier.cpp`? or just increment like `k_nearest_neighbors2.cpp`  if in the future someone implement a regression version?

Although the core algorithm is the same, when applied to a regression task the final algorithm has some differences from the one applied to a classification task. E.g: [KNeighborsRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html) and [KNeighborsClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html#sklearn.neighbors.KNeighborsClassifier)

**Comment by Panquesito7** [comment]
I think it's fine to keep the name as-is, not sure what the others think, though.

**Review by tjgurwara99 — APPROVED** [review]
